pkg/provisioner/ironic/ironic.go

```diff
@@ -1129,6 +1129,75 @@ func (p *ironicProvisioner) Adopt(force bool) (result provisioner.Result, err er
 	return
 }
 
+func (p *ironicProvisioner) startManualCleaning(ironicNode *nodes.Node) (result provisioner.Result, err error) {
+	var cleanSteps []nodes.CleanStep
+
+	// Start manual clean
+	if len(cleanSteps) != 0 {
+		p.log.Info("remove existing configuration and set new configuration", "steps", cleanSteps)
+		opts := nodes.ProvisionStateOpts{Target: nodes.TargetClean, CleanSteps: cleanSteps}
+		result, err = p.changeNodeProvisionState(ironicNode, opts)
+	}
+	return
+}
+
+// Prepare remove existing configuration and set new configuration
+func (p *ironicProvisioner) Prepare() (result provisioner.Result, err error) {
+	var ironicNode *nodes.Node
+
+	if ironicNode, err = p.findExistingHost(); err != nil {
+		return result, errors.Wrap(err, "could not find host to clean")
+	}
+	if ironicNode == nil {
+		return result, fmt.Errorf("no ironic node for host")
+	}
+
+	switch nodes.ProvisionState(ironicNode.ProvisionState) {
+	case nodes.Available, nodes.Active:
+		// Do Nothing (go to provisioning state)
+
+	case nodes.Manageable:
+		if p.host.Status.ManualCleaning {
+			// Manual clean finished
+			p.host.Status.ManualCleaning = false
+			return
+		}
+		result, err = p.startManualCleaning(ironicNode)
+		if result.Dirty == true && err == nil {
+			p.host.Status.ManualCleaning = true
+		}
+		return
+
+	case nodes.CleanFail:
+		if p.host.Status.ManualCleaning {
+			p.host.Status.ManualCleaning = false
+			result.Dirty = true
+			return
```

Comment:
> AIUI from #725 (comment) the change of Node state from manageable->cleaning is synchronous, so no danger on that front.

AFAIK record update is not synchronous (otherwise the API call will have to wait the db operation before answering). cc @dtantsur

Reply:
> the API call will have to wait the db operation before answering

That's the only sane way to do it, otherwise there's no way to know if your 'accepted' request will ever be acted upon or will just be lost without a trace. Writing the target state to the DB is not a particularly expensive operation.

As far as I can tell this is how it works - the [handler](https://opendev.org/openstack/ironic/src/branch/master/ironic/api/controllers/v1/node.py#L2107) [calls](https://opendev.org/openstack/ironic/src/branch/master/ironic/api/controllers/v1/node.py#L2216-L2218) the RPC [`update_node` method](https://opendev.org/openstack/ironic/src/branch/master/ironic/conductor/rpcapi.py#L233-L234), which [uses synchronous `call`](https://opendev.org/openstack/ironic/src/branch/master/ironic/conductor/rpcapi.py#L257-L258) (not asynchronous `cast`) to get the conductor to [save](https://opendev.org/openstack/ironic/src/branch/master/ironic/conductor/manager.py#L280) the node to the DB. So it's all synchronous.

AFAICT the only part in question is whether the provisioning state gets updated at the same time as we set the target state. I read Dmitry's comment as saying that since we are transitioning from a stable state (manageable) the current state has to be updated as part of that synchronous call.

Reply:
There are systems where such behavior could not be acceptable and where the DB tasks are considered expensive. Thanks for pointing out the Ironic's code - IIRC the observed behavior in #725 was that also the current state wasn't update synchronously, but I will recheck it. In any case probably it won't matter anymore if the flag approach will be dropped.

Reply:
A transition to a transient state is synchronous, i.e. after you issue a provisioning request you can expect the node to be in a new state on the next read (with an exception of a potentially quick enroll -> (manage) -> verifying -> (fail) -> enroll process).

Reply:
:+1: Thanks for clarifying @dtantsur